transports/webrtc-websys/src/connection.rs

```diff
@@ -0,0 +1,216 @@
+    /// Takes the RtcPeerConnection and DataChannelConfig and creates a pollable future
+    fn poll_create_data_channel(
+        &mut self,
+        cx: &mut Context,
+        config: DataChannelConfig,
+    ) -> Poll<Result<WebRTCStream, Error>> {
+        // Create Data Channel
+        // take the peer_connection and DataChannelConfig and create a pollable future
+        let mut dc = config.create_from(&self.peer_connection);
+        let channel = WebRTCStream::new(dc);
+        Poll::Ready(Ok(channel))
+    }
+
+    /// Polls the ondatachannel callback for incoming data channels.
+    ///
+    /// To poll for inbound WebRTCStreams, we need to poll for the ondatachannel callback
+    /// We only get that callback for inbound data channels on our connections.
+    /// This callback is converted to a future using CbFuture, which we can poll here
+    fn poll_ondatachannel(&mut self, cx: &mut Context) -> Poll<Result<WebRTCStream, Error>> {
+        // Poll the ondatachannel callback for incoming data channels
+        let dc = ready!(self.ondatachannel_fut.poll_unpin(cx));
+
+        // Create a WebRTCStream from the Data Channel
+        let channel = WebRTCStream::new(dc);
+        Poll::Ready(Ok(channel))
+    }
+
+    /// Closes the Peer Connection.
+    ///
+    /// This closes the data channels also and they will return an error
+    /// if they are used.
+    fn close_connection(&mut self) {
+        if !self.closed {
+            self.peer_connection.close();
+            self.closed = true;
+        }
+    }
+}
+
+pub(crate) async fn register_data_channel(
+    conn: &RtcPeerConnection,
+    config: &DataChannelConfig,
+) -> RtcDataChannel {
+    // peer_connection.set_ondatachannel is callback based
+    // but we need a Future we can poll
+    // so we convert this callback into a Future by using [CbFuture]
+
+    // 1. create the ondatachannel callbackFuture
+    // 2. build the channel with the DataChannelConfig
+    // 3. await the ondatachannel callbackFutures
+    // 4. Now we have a ready DataChannel
+    let ondatachannel_fut = CbFuture::new();
+    let cback_clone = ondatachannel_fut.clone();
+
+    debug!("register_data_channel");
+    // set up callback and futures
+    // set_onopen callback to wake the Rust Future
+    let ondatachannel_callback = Closure::<dyn FnMut(_)>::new(move |ev: RtcDataChannelEvent| {
+        let dc2 = ev.channel();
+        debug!("ondatachannel! Label (if any): {:?}", dc2.label());
+
+        cback_clone.publish(dc2);
+    });
+
+    conn.set_ondatachannel(Some(ondatachannel_callback.as_ref().unchecked_ref()));
+
+    let _dc = config.create_from(conn);
+
+    ondatachannel_fut.await
+}
+
+impl Drop for ConnectionInner {
+    fn drop(&mut self) {
+        self.close_connection();
+    }
+}
+
+/// WebRTC native multiplexing
+/// Allows users to open substreams
+impl StreamMuxer for Connection {
+    type Substream = WebRTCStream; // A Substream of a WebRTC PeerConnection is a Data Channel
```

Comment:
> I was confused by this too, so I'm glad. The `Substream` in `StreamMuxer` affects other crates though (pretty widespread, mplex, etc.), so perhaps we make that change separately through a different PR?

Sorry, I was only referring to the name of the type `WebRTCStream`. The associated type should stay as is for now and be changed in a separate PR. I didn't want to imply that you should rename the associated type :)

Reply:
Oh so you are saying rename `WebRTCStream` to `Stream`, ok

Reply:
I did find the naming it easier to differentiate between all the `Stream`s though (futures, muxer, which stream is which)

Reply:
> I did find the naming it easier to differentiate between all the `Stream`s though (futures, muxer, which stream is which)

I get you! It is a trade-off and our assumption is that it is better for our users if they can say `use libp2p::webrtc_websys` and then refer to types as `webrtc_websys::Transport` and `webrtc_websys::Stream` instead of having to import all types because they are already fully-qualified.

It makes life a bit harder for us but I personally usually search for types via the module name too. CLion supports that quite well, I am not sure about other IDEs. I think @mxinden uses emacs and rust-analyzer but I've not heard any complaints from him so far. Perhaps he can share some workflows too!

Reply:
Thanks for the explanations gents, I've made the changes to `Stream` as discussed